chainerx_cc/chainerx/routines/normalization.h

```diff
@@ -1,15 +1,120 @@
 #pragma once
 
+#include <array>
+#include <memory>
+#include <tuple>
+
 #include <nonstd/optional.hpp>
 
 #include "chainerx/array.h"
 #include "chainerx/axes.h"
 #include "chainerx/constant.h"
+#include "chainerx/op.h"
 #include "chainerx/scalar.h"
 #include "chainerx/stack_vector.h"
 
 namespace chainerx {
 
+// Intermediate results from `BatchNormOp::Call` can be stored in this construct and be reused in `BatchNormGradOp::Call`.
+// The objects to store may vary depending on backend so each backend should derive this class to define the actual set of intermediate
+// results.
+class BatchNormState {
+public:
+    virtual ~BatchNormState() = default;
+};
+
+class BatchNormOp : public Op {
+public:
+    static const char* name() { return "BatchNormForward"; }
+
+    // The returned state should be a `nullptr` if `return_state` is `false`.
+    virtual std::tuple<Array, std::shared_ptr<BatchNormState>> Call(
+            const Array& x,
+            const Array& gamma,
+            const Array& beta,
+            const Array& running_mean,
+            const Array& running_var,
+            Scalar eps,
+            Scalar decay,
+            const Axes& axis,
+            bool return_state,
+            const nonstd::optional<Array>& out = nonstd::nullopt) = 0;
```

Comment:
Are default values needed?
They're not meant to be directly called by users.

Reply:
Fixed. Looks quite verbose for backward though.